### PR TITLE
cache special_read_file content

### PR DESCRIPTION
**Pull request by songgao**

### CODE REVIEW

File: libfuse/special_read_file.go [hunk 1]
```diff
@@ -16,26 +17,51 @@ import (
...
 	read func(context.Context) ([]byte, time.Time, error)
+
+	dataLock   sync.Mutex
+	data       []byte
+	ts         time.Time
+	dataExpire time.Time
 }
 
+// Have a low non-zero value for Valid to avoid being swamped with requests,
+// while still keeping the size up to date.
```
Comment (akalin-keybase) [inline libfuse/special_read_file.go]: Add a comment about the docker poll time being 1s, so making this 0.5s.

File: libfuse/special_read_file.go [hunk 2]
```diff
@@ -16,26 +17,51 @@ import (
...
+const specialReadFileValidDuration = time.Second / 2
+
 var _ fs.Node = (*SpecialReadFile)(nil)
 
+func (f *SpecialReadFile) cachedDataExpiredLocked() bool {
```
Comment (akalin-keybase) [inline libfuse/special_read_file.go]: this function is unused

File: libfuse/special_read_file.go [hunk 3]
```diff
@@ -16,26 +17,51 @@ import (
...
+	return time.Now().After(f.dataExpire)
+}
+
+func (f *SpecialReadFile) refreshCachedDataIfNeeded(ctx context.Context) (err error) {
+	f.dataLock.Lock()
+	defer f.dataLock.Unlock()
+	if time.Now().Before(f.dataExpire) {
```
Comment (akalin-keybase) [inline libfuse/special_read_file.go]: prefer storing time.Now() in a local variable and using it here and below

File: libfuse/special_read_file.go [hunk 4]
```diff
@@ -16,26 +17,51 @@ import (
 // a function.
 type SpecialReadFile struct {
 	read func(context.Context) ([]byte, time.Time, error)
+
+	dataLock   sync.Mutex
```
Comment (akalin-keybase) [inline libfuse/special_read_file.go]: There's an equivalent type in libdokan, too. Perhaps fix that one also? Up to you.

**Comment by songgao** [comment]
Ugh this breaks DSL tests. I guess we should leave it the old way ...

**Comment by akalin-keybase** [comment]
This PR is giving me deja vu...I feel like I tried this once before and maybe came to the same conclusion. :sweat_smile:

**Comment by songgao** [comment]
😂

**Pull request closed without merging**